### PR TITLE
WIP: first draft for term1.0 module

### PR DESCRIPTION
**Pull request by majewsky**
This is very much work-in-progress, but if you see anything unclear or missing, feel free to comment.

### CODE REVIEW

File: spec/term/1.0.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+TODO disabled by default; when enabled, server MUST make all keyboard input available on stdin immediately instead of waiting for a full line
+
+### 3.5. The `term1.input-echo` property
+
+TODO enabled by default; when enabled, printable characters entered by the user are reflected on the terminal; when disabled, may only be re-enabled on the server connection where it was disabled; will be re-enabled automatically when the server connection disabling it is closed
+
+### 3.6. The `term1.output-protected` property
+
+TODO disabled by default; when enabled, server MUST disregard all unprintable characters and control characters (incl. \<ESC\>) except for CR/LF/HT on any standard input; may only be disabled on the server connection where it was enabled; will be disabled automatically when the server connection enabling it is closed (the purpose of this is to prevent exploits where carefully crafted shell scripts appear benign when printed with `cat(1)`, but contain malicious code which is hidden with backspace chars and VT100 control sequences)
```
Comment (majewsky) [inline spec/term/1.0.md:105]: > may be disabled only on othe server connection where it was enabled

This creates a race: When stdout ends with a control sequence, whether it is interpreted depends on whether the terminal drains stdout first, or reads and executes the `(core1.set term1.output-protected false)` instruction first.

I think it's inevitable that we come up with some sort of equivalent to `TCSA{NOW,DRAIN,FLUSH}` (see `man 3 tcsetattr`).

File: spec/term/1.0.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+*Rationale:* This rule allows terminals to be backwards-compatible with legacy control sequences, such as those defined in VT100 aka ECMA-48 aka ANSI X3.64 aka ISO/IEC 6429.
+
+Characters that are in the general category Co ("Other, private use") MAY be exempted from the treatment described above if the characters in question have a graphic representation in the font that the terminal uses to display its character array.
+
+Finally, the server MUST discard each U+000A (new line) character that immediately succeeds a U+000D (carriage return) character, and then convert each U+000D character into a U+000A character.
+
+*Rationale:* This rule ensures consistent line break handling across different platforms, i.e., CR+NL = CR = NL.
+
+The remaining string of Unicode characters SHALL then be written into the character array starting from the position pointed to by the cursor, using the rules for normalization, writing direction, line breaking and handling of bidirectional text as laid out in the Unicode Standard, as published by the [Unicode consortium](https://www.unicode.org).
```
Comment (majewsky) [inline spec/term/1.0.md:64]: Note to self: Should check how the Unicode standard is referenced in other standards that concern text layouting, e.g. HTML/CSS.

**Review by majewsky — COMMENTED** [review]
We will probably need to go back to the whiteboard here, and also have a look at prior art (such as that linked in the survey submissions).

File: spec/term/1.0.md (hunk 1 above)
Comment (majewsky) [reply spec/term/1.0.md:105]: Correction: There is no race. `core1.set` is answered with `core1.pub`, which the application can wait for before starting to write the protected output data.

File: spec/term/1.0.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+The character array is made up of **columns** of characters along the horizontal dimension, and of **rows** of characters along the vertical dimension.
+The number of columns and rows in the character array is called the terminal's **width** and **scrollback height**, respectively.
+
+Columns are numbered from left to right, and rows are numbered from top to bottom, in both cases starting with zero.
+A position in the array is addressed by a pair of the index of its row and the index of its column.
+For example, the position `(0,0)` refers to the character at the top left corner of the array.
+
+For graphical terminals, the character array MAY be much larger than what can be displayed on screen at once.
+In this case, the character array itself is called the terminal's **scrollback**, and the range of currently visible columns is called the terminal's **viewport**.
+The number of columns in the viewport is called the terminal's **viewport height**.
```
Comment (majewsky) [inline spec/term/1.0.md]: We need a viewport width, too.

File: spec/term/1.0.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+*Rationale:* The term "terminal height" is not defined to avoid confusion. We always refer explicitly to either the scrollback height or the viewport height.
+
+### 1.2. Standard output
+
+When a client writes a byte string into its standard output, the server SHALL assume that this byte string is encoded in [UTF-8](https://tools.ietf.org/html/rfc3629), and decode it into a string of Unicode characters.
+
+When the `term1.output-protected` property is true, the server MUST discard all characters from this string that are in the general categories Cc ("Other, control"), Cf ("Other, format"), Cs ("Other, surrogate"), Co ("Other, private use") and Cn ("Other, not assigned"), except for U+0009 (Horizontal Tab), U+000A (New Line) and U+000D (Carriage Return).
```
Comment (majewsky) [inline spec/term/1.0.md:47]: Also, U+000C (Form Feed), to make line printers happy.

**Review by laerling — COMMENTED** [review]
(all changes discussed in person)

File: spec/term/1.0.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+When the `term1.output-protected` property is false, the server MAY alternatively:
+
+- choose a printable representation for each such character and replace the character with that representation before continuing, or
+- interpret the character as starting a control sequence if the character is U+001B (Escape).
+
+When a control character is interpreted as starting a control sequence, the entire control sequence SHALL be removed from the string before processing as described in this section continues, and the remaining string MAY be modified depending on the semantics of the control sequence.
```
Comment (majewsky) [inline spec/term/1.0.md:54]: Need to add a requirement that ISO 2022 sequences MUST NOT be interpreted under any circumstances. *Rationale:* It's the 21st century.

**Comment by majewsky** [comment]
Discarding this in favor of #32.

**Pull request closed without merging**